File: jdk/test/sun/security/ssl/StatusStapling/java.base/sun/security/ssl/TestUtils.java
/*
 * Copyright (c) 2015, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package sun.security.ssl;

import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.Map;

public class TestUtils {

    // private constructor to prevent instantiation for this utility class
    private TestUtils() {
        throw new AssertionError();
    }

    public static void runTests(Map<String, TestCase> testList) {
        int testNo = 0;
        int numberFailed = 0;
        Map.Entry<Boolean, String> result;

        System.out.println("============ Tests ============");
        for (String testName : testList.keySet()) {
            System.out.println("Test " + ++testNo + ": " + testName);
            result = testList.get(testName).runTest();
            System.out.print("Result: " + (result.getKey() ? "PASS" : "FAIL"));
            System.out.println(" " +
                    (result.getValue() != null ? result.getValue() : ""));
            System.out.println("-------------------------------------------");
            if (!result.getKey()) {
                numberFailed++;
            }
        }

        System.out.println("End Results: " + (testList.size() - numberFailed) +
                " Passed" + ", " + numberFailed + " Failed.");
        if (numberFailed > 0) {
            throw new RuntimeException(
                    "One or more tests failed, see test output for details");
        }
    }

    public static void dumpBytes(byte[] data) {
        dumpBytes(ByteBuffer.wrap(data));
    }

    public static void dumpBytes(ByteBuffer data) {
        int i = 0;

        data.mark();
        while (data.hasRemaining()) {
            if (i % 16 == 0 && i != 0) {
                System.out.print("\n");
            }
            System.out.print(String.format("%02X ", data.get()));
            i++;
        }
        System.out.print("\n");
        data.reset();
    }

    public static void valueCheck(byte[] array1, byte[] array2) {
        if (!Arrays.equals(array1, array2)) {
            throw new RuntimeException("Array mismatch");
        }
    }

    // Compares a range of bytes at specific offsets in each array
    public static void valueCheck(byte[] array1, byte[] array2, int skip1,
            int skip2, int length) {
        ByteBuffer buf1 = ByteBuffer.wrap(array1);
        ByteBuffer buf2 = ByteBuffer.wrap(array2);

        // Skip past however many bytes are requested in both buffers
        buf1.position(skip1);
        buf2.position(skip2);

        // Reset the limits on each buffer to account for the length
        buf1.limit(buf1.position() + length);
        buf2.limit(buf2.position() + length);

        if (!buf1.equals(buf2)) {
            throw new RuntimeException("Array range mismatch");
        }
    }

    // Concatenate 1 or more arrays
    public static byte[] gatherBuffers(byte[]... arrays) {
        int totalLength = 0;
        for (byte[] ar : arrays) {
            totalLength += ar != null ? ar.length : 0;
        }

        byte[] resultBuf = new byte[totalLength];
        int offset = 0;
        for (byte[] ar : arrays) {
            if (ar != null) {
                System.arraycopy(ar, 0, resultBuf, offset, ar.length);
                offset += ar.length;
            }
        }
        return resultBuf;
    }
}

File: test/hotspot/gtest/testutils.cpp
/*
 * Copyright (c) 2021, 2023 SAP SE. All rights reserved.
 * Copyright (c) 2021, 2025, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

#include "runtime/os.hpp"
#include "utilities/align.hpp"
#include "utilities/ostream.hpp"

#include "testutils.hpp"
#include "unittest.hpp"

#include "os_bsd.hpp"
#include <string.h>

// Note: these could be made more suitable for covering large ranges (e.g. just mark one byte per page).

void GtestUtils::mark_range_with(void* p, size_t s, uint8_t mark) {
  if (p != nullptr && s > 0) {
    ::memset(p, mark, s);
  }
}

bool GtestUtils::is_range_marked(const void* p, size_t s, uint8_t expected) {
  if (p == nullptr || s == 0) {
    return true;
  }

  const char* first_wrong = nullptr;
  char* p2 = (char*)p;
  const char* const end = p2 + s;
  while (p2 < end) {
    if (*p2 != (char)expected) {
      first_wrong = p2;
      break;
    }
    p2 ++;
  }

  if (first_wrong != nullptr) {
    tty->print_cr("check_range [" PTR_FORMAT ".." PTR_FORMAT "), 0x%X, : wrong pattern around " PTR_FORMAT,
                  p2i(p), p2i(p) + s, expected, p2i(first_wrong));
    // Note: We deliberately print the surroundings too without bounds check. Might be interesting,
    // and os::print_hex_dump uses SafeFetch, so this is fine without bounds checks.
    os::print_hex_dump(tty, (address)(align_down(p2, 0x10) - 0x10),
                            (address)(align_up(end, 0x10) + 0x10), 1);
  }

  return first_wrong == nullptr;
}

#if APPLE_MEMORY_TAGGING_AVAILABLE
bool GtestUtils::is_memory_tagged_as_java(void* addr, size_t size) {
  // Use mach_vm_region with extended info to get the user_tag
  mach_vm_address_t address = (mach_vm_address_t)addr;
  mach_vm_size_t region_size = 0;
  vm_region_extended_info_data_t extended_info;
  mach_msg_type_number_t info_count = VM_REGION_EXTENDED_INFO_COUNT;
  mach_port_t object_name = MACH_PORT_NULL;

  kern_return_t kr = mach_vm_region(mach_task_self(),
                                    &address,
                                    &region_size,
                                    VM_REGION_EXTENDED_INFO,
                                    (vm_region_info_t)&extended_info,
                                    &info_count,
                                    &object_name);

  if (kr != KERN_SUCCESS) {
    return false;
  }

  // Check if the memory region covers our allocation and has the correct tag
  if (address <= (mach_vm_address_t)addr &&
      (address + region_size) >= ((mach_vm_address_t)addr + size)) {
    return extended_info.user_tag == VM_MEMORY_JAVA;
  }

  return false;
}
#endif // APPLE_MEMORY_TAGGING_AVAILABLE
